### DISCUSSION TITLE
Uglify Does Not Work in TW 5.2.0

### DISCUSSION THREAD

**Discussion by kookma** [TiddlyWiki/TiddlyWiki5#6043] [Q&A] [answered]
Uglify a plugin lets to minify the javascripts in TW does not work in TW 5.2.0.

Uglify is very helpful to reduce the size of single TW.html from around 2.364MB to 1.442MB in TW 5.1.23
This is helpful for us when distributing TW through emails and collect students TW for a term paper or course project!

I had some discussion with @flibbles but he was not able to address the issue! See [2]

I would like to know, is it possible to have an option to install compressed or uncompressed core in our wiki like any other plugins?

Ref:

1. https://groups.google.com/g/tiddlywiki/c/kF-OGmkyFfo/m/b272I2rtAwAJ
2. https://github.com/flibbles/tw5-uglify/issues/2

**Comment by flibbles** [comment]
So for anyone who's interested, the reason Uglify doesn't work in V5.2.0 is because then entire conversion of the tiddler base down to a file to save happens in the new jsonwidget:
https://github.com/Jermolene/TiddlyWiki5/blob/575c2335975cc6cc960d2061ba8fdfa726ed8618/core/modules/widgets/jsontiddler.js#L35-L49
Everything happens here, and there is no way for me to interject the ability to modify the content of a tiddler before it gets written to file. If people are interested in modifying this to allow such a thing to happen, I'd be happy to migrate Uglify.

**Comment by pmario** [comment]
What, if you change your concept a bit. Let's assume, there is an "uglify edition", Which is empty.html plus the plugin.

 - You open the edition
 - Drag and drop import the wiki, that should be uglified
 - The import mechanism has several new hooks in 5.2.0 - They will allow you to "touch every" tiddler in the different stages of the import procedure
 - There is a user facing UI, that will allow the user to decide which elements should be imported.
 - Once the code tiddlers are uglified, the default save mechanism will be enough. 
 - The saver configuration can save the whole thing without the plugin, since it isn't needed anymore.

What do you think?

**Reply by flibbles** [reply]
So you'd have them drag their own project onto the Uglify website and then download the results? I suppose that would work for standalone projects. Though I'm not sure how this answers the true reason I wrote Uglify, which was to discourage people like tobibeer and felixhayashi from publishing their plugins uglified, and instead leave it up to the users whether they wanted uglified projects. In other words, how to use uglified on Node.JS. I'd still love to have Node.JS have directories full of uncompressed plugins which it compresses before sending to the browser.

**Reply by pmario** [reply]
hmmm, OK. 

There should be a possibility, to add a hook in line 40 eg: 

```
 if(this.attTiddler) { 
 	var tiddler = this.wiki.getTiddler(this.attTiddler); 
 	if(tiddler) { 
 		fields = tiddler.getFieldStrings({exclude: this.attExclude.split(" ")}); 
                fields = $tw.hooks.invokeHook("th-get-json-fields",fields);
 	} 
 } 
```

@Jermolene what do you think?

**Reply by flibbles** [reply]
I could get in there, or I could just take getFieldStrings and monkey patch it. I'd rather prefer your way.

Though doesn't what you're showing here cause fields to become undefined again if there aren't any hooks?

**Reply by pmario** [reply]
No the  invokeHook() function returns the unmodified first argument if there is no hook.

**Reply by pmario** [reply]
see: https://github.com/Jermolene/TiddlyWiki5/blob/fb4d77ef46e23c23f4e46c13f4a529aa29a27dcd/boot/boot.js#L2569-L2577

**Reply by flibbles** [reply]
Ooh. Smart. Then yeah, this would provide a trivial way for me to make it work in both cases.

**Reply by saqimtiaz** [reply]
@flibbles I support the idea of creating a PR to add a hook.

**Reply by Jermolene** [reply]
A hook may not be necessary in this case: an alternative would be to override the templates `$:/core/templates/html-json-tiddler` and `$:/core/templates/html-json-skinny-tiddler` to use a different custom widget.

**Reply by flibbles** [reply]
I suppose that works too. Makes it less clean for me though. I'll need to copy-past the jsontiddler and keep it up to date if it's ever changed in the future.

Is your concern that this hook wouldn't have much use-case outside of Uglify's particular circumstances?

**Reply by Jermolene** [reply]
> Is your concern that this hook wouldn't have much use-case outside of Uglify's particular circumstances?

I think of hooks as a last report: they are very brittle, and tend to constrain the internal implementation of primitives. Also, over time we hope to make more of TiddlyWiki's logic be expressed in wikitext (eg most of what the navigator widget does), which will mean that ultimately we will have to drop support for many of the existing hooks.

**Reply by flibbles** [reply]
I see. Would you be against separating out a bit of the logic in jsonTiddler so the field fetching logic is in its own method? A la...
```javascript
JSONTiddlerWidget.prototype.getTiddlerFields = function() {
	var fields = {};
	if(this.attTiddler) {
		var tiddler = this.wiki.getTiddler(this.attTiddler);
		if(tiddler) {
			fields = tiddler.getFieldStrings({exclude: this.attExclude.split(" ")});
		}
	}
	return fields;
};
```
This way, anyone (me) wanting to change that behavior need only inherit from JSONWidget and override that one method? Or is that still constraining the design, and you'd still prefer a custom widget?

**Reply by Jermolene** [reply]
No that's fine, it makes sense to make JSONTiddlerWidget more easily subclassable.

**Reply by flibbles** [reply]
Cool beans. I'll make a PR for it when I have a moment.

**Reply by pmario** [reply]
@flibbles ... Just to be sure, you know of the `module-type: widget-subclass` method for subclassing existing widgets.

**Reply by flibbles** [reply]
I did not. It sounds relevant though. I'll look into it.

**Reply by pmario** [reply]
It's a way to create your own widgets, that are based on core widgets, without the need to monkeypatch the core widget. See: https://github.com/wikilabs/plugins/blob/master/wikilabs/custom-markup/tiddlers/widgets/wl-keyboard.js where I did create a custom-widget. 

It has a plugin specific name and only contains the function that has been overwritten .. So it should be easy for you to create your own, modified jsontiddler-widget. Where only a minimum code is changed. ...

**Reply by pmario** [reply]
also see: https://tiddlywiki.com/dev/#WidgetSubclassingMechanism

**Reply by flibbles** [reply]
Ah, whoops. Doesn't look like it works for me.
https://github.com/Jermolene/TiddlyWiki5/blob/0be39cfbc2829dbd8637f35110b33c4bd2416dc3/core/modules/widgets/widget.js#L56-L59
If I want to maintain support for V5.1.*, I can't use this, or else my subclass will cause TiddlyWiki to throw an exception.

**Reply by pmario** [reply]
>If I want to maintain support for V5.1.*, I can't use this, or else my subclass will cause TiddlyWiki to throw an exception.

hmmm **confused**, TW v5.1.x doesn't use this format to store tiddler content. ...

**Reply by flibbles** [reply]
It doesn't, true. Writes them out using the <$view> widget, and I'll have to keep that either way since that's still how it writes boot.js and boot-prefix.js. However, if uglify is given an old version of tiddlywiki, then even though it won't have the JSONTiddlerWidget, having a subclass of it in my widget will cause an exception.

Although now that I think about it, I can just do something like:
```javascript
if (Widget.prototype.jsontiddler) {
    exports.baseClass = "jsontiddler";
}
```
And then my subclass will get ignored if jsontiddler isn't there.

**Comment by flibbles** [comment, marked as answer]
All right. A new version of Uglify is out, which will work with Tiddlywiki V5.2.0 just as soon as #6057 has been merged.

@kookma, I did this for you, buddy. I'm not sure anyone else out there uses Uglify besides you and me.

**Reply by saqimtiaz** [reply]
@flibbles thank you. I have two colleagues who wouldn't be using TW at all if it wasn't for Uglify, due to bandwidth constraints.

**Reply by kookma** [reply]
> All right. A new version of Uglify is out, which will work with Tiddlywiki V5.2.0 just as soon as #6057 has been merged.
> 
> @kookma, I did this for you, buddy. I'm not sure anyone else out there uses Uglify besides you and me.

Much appreciated @flibbles 
Beside the two colleagues of Saq, I have bunch of students all uses the uglify version! None of us know JS  ;-)

**Reply by kookma** [reply]
> All right. A new version of Uglify is out, which will work with Tiddlywiki V5.2.0 just as soon as #6057 has been merged.

@flibbles - with regard to the new TW update with https://github.com/Jermolene/TiddlyWiki5/pull/6057 merged, is there any need to have an Uglify update too? It seems the current version does not work!

**Reply by flibbles** [reply]
Seems to be working for me. Let's take this discussion to https://github.com/flibbles/tw5-uglify/issues/2 since this no longer concerns core TiddlyWiki developers.
